### ISSUE TITLE
Add backpressure to decoder scheduler

### ISSUE THREAD

**Issue by westonpace** [lancedb/lance#1957]
I'm not certain if this goes in the decode scheduler or the file scheduler but I think it is in the decoder scheduler.  The purpose of this queue is to handle the case where the compute threads are falling behind the I/O.  Though unlikely, this could happen, for example, on a busy system with fast storage.  If we don't pause the scheduler we will keep scheduling I/O and the I/O scheduler will keep delivering data and memory will grow.

The backpressure should be configured as "X rows of backpressure" and should probably default to something like <SOME_HEURISTIC * BATCH_SIZE>.

We need to be careful to ensure that the backpressure is only increased when enough rows to deliver a batch are accumulated.  E.g. if we increased the backpressure one page at a time and there were 10,000 columns and so at least 10,000 pages needed to make a row and we then scheduled 5,000 of these pages then we can't block because the decoder will never be able to generate a batch and relieve backpressure.

Indirect follow-up requests will need to be able to bypass the backpressure to avoid deadlock.  For example, imagine that the scheduler schedules a single row targeting a page of string data.  The scheduler will continue as soon as the first request for the data offsets is launched.  Then, the scheduler might schedule 100 more pages and fill up the backpressure queue.  Then, the string offsets will finish loading and a follow-up task will be submitted to fetch the string data.  If this follow-up task is blocked by the backpressure queue then the decoder (which decodes in order) will never get the string data and halt.  The backpressure will never be relieved because the decoder is blocked.

**Comment by westonpace** [comment]
Closed by #2683
